dsl/src/commonMain/kotlin/com/freeletics/flowredux/dsl/flow/WhileInState.kt

```diff
@@ -0,0 +1,46 @@
+package com.freeletics.flowredux.dsl.flow
+
+import com.freeletics.flowredux.GetState
+import com.freeletics.flowredux.dsl.internal.Action
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.launch
+
+@ExperimentalCoroutinesApi
+internal fun <S, A> Flow<Action<S, A>>.whileInState(
+    isInState: (S) -> Boolean,
+    getState: GetState<S>,
+    transform: (Flow<Action<S, A>>) -> Flow<Action<S, A>>
+) = channelFlow {
```

Comment:
Why do we need a ChannelFlow here?. wouldn't it work with just a regular Flow as well?

Reply:
The reason is that we emit/send from within the `launch`. Just using `flow { coroutineScope { ... } }` violates the context preservation requirement of flow and crashes, you can see that in the commit from before the force push